internal/fleet/integration_policy/secrets.go

```diff
@@ -0,0 +1,165 @@
+package integration_policy
+
+import (
+	"context"
+	"encoding/json"
+
+	fleetapi "github.com/elastic/terraform-provider-elasticstack/generated/fleet"
+	"github.com/elastic/terraform-provider-elasticstack/internal/utils"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+)
+
+// The secret store is a map of policy secret reference IDs to the
+// original value at time of creation. By replacing the ref when
+// marshaling the state back to Terraform, we can prevent resource
+// drift.
+type secretStore map[string]any
+
+// newSecretStore creates a new secretStore from the resource privateData.
+func newSecretStore(ctx context.Context, private privateData) (store secretStore, diags diag.Diagnostics) {
+	bytes, diags := private.GetKey(ctx, "secrets")
+	if diags != nil {
+		return
+	}
+	if bytes == nil {
+		bytes = []byte("{}")
+	}
+
+	err := json.Unmarshal(bytes, &store)
+	if err != nil {
+		diags.AddError("could not unmarshal secret store", err.Error())
+		return
+	}
+
+	return
+}
+
+// Save marshals the secretStore back to the provider.
+func (s secretStore) Save(ctx context.Context, private privateData) (diags diag.Diagnostics) {
+	bytes, err := json.Marshal(s)
+	if err != nil {
+		diags.AddError("could not marshal secret store", err.Error())
+		return
+	}
+
+	return private.SetKey(ctx, "secrets", bytes)
+}
+
+// pruneRefsFromResponse removes any saved secret refs not present in the API response.
+func pruneRefsFromResponse(resp *fleetapi.PackagePolicy, secrets secretStore) {
+	refs := make(map[string]any)
+	for _, r := range utils.Deref(resp.SecretReferences) {
+		refs[*r.Id] = nil
+	}
+
+	for id := range secrets {
+		if _, ok := refs[id]; !ok {
+			delete(secrets, id)
+		}
+	}
+}
+
+// handleRespSecrets extracts the wrapped value from each response var, then
+// replaces any secret refs with the original value from secrets if available.
+func handleRespSecrets(resp *fleetapi.PackagePolicy, secrets secretStore) {
+	handleVars := func(vars map[string]any) {
+		for key, val := range vars {
+			if mval, ok := val.(map[string]any); ok {
+				if wrapped, ok := mval["value"]; ok {
+					vars[key] = wrapped
+					val = wrapped
+				} else {
+					// Don't keep null (missing) values
+					delete(vars, key)
+					continue
+				}
+
+				if mval, ok := val.(map[string]any); ok {
+					if v, ok := mval["isSecretRef"]; ok && v == true {
+						refID := mval["id"].(string)
+						if original, ok := secrets[refID]; ok {
+							vars[key] = original
+						}
+					}
+				}
+			}
+		}
+	}
+
+	handleVars(utils.Deref(resp.Vars))
```

Comment:
fwiw, without using the x-go-remove-unnecessary-pointers flag, you have to sprinkle these deref methods everywhere. Being able to range a nil map is nice.